### PR TITLE
Improve time functions

### PR DESCRIPTION
**Pull request by maraino**
### Description

This PR improves the functions added in https://github.com/smallstep/crypto/pull/534 with some usability changes. 

* `toTime 1719970524` returns a `time.Time` in UTC from a time or a Unix epoch.
* `formatTime (now)`: from a time or Unix epoch returns a time string using RFC3339 format UTC.
* `parseTime` and `mustParseTime`:
  * `parseTime`: returns the current `time.Time` in UTC.
  * `parseTime "2024-07-03T01:30:39Z"`: parses the given time using RFC3339.
  * `parseTime "time.UnixDate" "Tue Jul  2 18:31:04 PDT 2024"`: parses a time using a given format.
  * `parseTime "time.UnixDate" "Tue Jul  2 18:31:04 PDT 2024" "America/Los_Angeles"`: parses a time with the given format and time zone.
* `toTimeLayout`  returns a layout format from the given name.
  * `toTimeLayout "RFC3339"` returns `2006-01-02T15:04:05Z07:00`
  * `toTimeLayout "time.UnixDate"` returns `Mon Jan _2 15:04:05 MST 2006`

### CODE REVIEW

File: internal/templates/funcmap.go [hunk 1]
```diff
@@ -10,12 +11,16 @@ import (
...
-// RFC3339 in UTC. The "fail" function sets the provided message, so that
-// template errors are reported directly to the template without having the
-// wrapper that text/template adds.
+// The "formatTime" function receives a time or a Unix epoch and formats it to
+// RFC3339 in UTC. The "toTime" and "mustToTime" functions parses a formatted
+// string and returns the time value it represents. The "toLayout" function
+// converts strings like "time.RFC3339" or "time.UnixDate" to the actual layout
+// represented by the Go constant with the same name. The "fail" function sets
+// the provided message, so that template errors are reported directly to the
+// template without having the wrapper that text/template adds.
```
[the remark this answers is not in the document]
Comment (maraino) [reply internal/templates/funcmap.go]: Fixed with ee3ab80

File: internal/templates/funcmap.go [hunk 2]
```diff
@@ -10,12 +11,16 @@ import (
 )
 
 // GetFuncMap returns the list of functions provided by sprig. It adds the
-// function "toTime" and changes the function "fail".
+// function "formatTime", "toTime", "mustToTime", and changes the function
+// "fail".
```
[the remark this answers is not in the document]
Comment (maraino) [reply internal/templates/funcmap.go]: Fixed with ee3ab80

File: internal/templates/funcmap.go [hunk 3]
```diff
@@ -27,11 +49,15 @@ func GetFuncMap(failMessage *string) template.FuncMap {
 		*failMessage = msg
 		return "", errors.New(msg)
 	}
+	m["formatTime"] = formatTime
 	m["toTime"] = toTime
+	m["parseTime"] = parseTime
+	m["mustParseTime"] = mustParseTime
+	m["toLayout"] = toLayout
```
Comment (hslatman) [inline internal/templates/funcmap.go:56]: The functionality of the function is specific to times/dates, and the other new functions have this in their name. Maybe it should be called toTimeLayout / toDateLayout (or support both)?

File: internal/templates/funcmap.go [hunk 4]
```diff
@@ -10,12 +11,33 @@ import (
 )
 
 // GetFuncMap returns the list of functions provided by sprig. It adds the
-// function "toTime" and changes the function "fail".
+// functions "formatTime", "toTime", "mustToTime", and changes the function
+// "fail".
```
Comment (hslatman) [inline internal/templates/funcmap.go:15]: Might want to add `toLayout` (or its updated name(s)) here

File: internal/templates/funcmap.go [hunk 5]
```diff
@@ -10,12 +11,33 @@ import (
...
-// The "toTime" function receives a time or a Unix epoch and formats it to
-// RFC3339 in UTC. The "fail" function sets the provided message, so that
-// template errors are reported directly to the template without having the
-// wrapper that text/template adds.
+// The "toTime" function receives a time or a Unix epoch and returns a time.Time
+// in UTC. The "formatTime" function uses "toTime" and formats the resulting
+// time using RFC3339. The functions "parseTime" and "mustParseTime" parse a
+// string and return the time.Time it represents. The "toLayout" function
+// converts strings like "time.RFC3339" or "time.UnixDate" to the actual layout
+// represented by the Go constant with the same nameß. The "fail" function sets
```
Comment (hslatman) [inline internal/templates/funcmap.go:22]:
```suggestion
// represented by the Go constant with the same name. The "fail" function sets
```

File: internal/templates/funcmap.go (hunk 3 above)
Comment (tashian) [reply internal/templates/funcmap.go:56]: Could be `toTimeLayout "RFC8222"`
instead of `toLayout "time.RFC8222"`?

Comment (maraino) [reply internal/templates/funcmap.go]: Fixed with a23cb16

File: internal/templates/funcmap.go (hunk 4 above)
Comment (maraino) [reply internal/templates/funcmap.go]: Fixed with a23cb16

File: internal/templates/funcmap.go [hunk 6]
```diff
@@ -53,5 +79,81 @@ func toTime(v any) string {
...
+		return time.RFC850
+	case "RFC1123":
+		return time.RFC1123
+	case "RFC1123Z":
+		return time.RFC1123Z
+	case "RFC3339":
+		return time.RFC3339
+	case "RFC3339Nano":
+		return time.RFC3339Nano
+	// From the ones bellow, only time.DateTime will parse a complete date.
```
Comment (hslatman) [inline internal/templates/funcmap.go:139]:
```suggestion
	// From the ones below, only time.DateTime will parse a complete date.
```

File: internal/templates/funcmap.go [hunk 7]
```diff
@@ -53,5 +79,81 @@ func toTime(v any) string {
...
+		}
+		return time.ParseInLocation(layout, v[1], loc)
+	default:
+		return time.Time{}, errors.New("unsupported number of parameters")
+	}
+}
+
+func toTimeLayout(fmt string) string {
+	switch strings.TrimPrefix(fmt, "time.") {
+	case "Layout":
```
Comment (hslatman) [inline internal/templates/funcmap.go:117]: Just for convenience for the user, we could allow any case, but up to you

**Review by hslatman — COMMENTED** [review]
Minor typo; potential change. But otherwise good 😄

File: internal/templates/funcmap.go (hunk 6 above)
Comment (maraino) [reply internal/templates/funcmap.go]: Fixed with 329b3f1

File: internal/templates/funcmap.go (hunk 7 above)
Comment (maraino) [reply internal/templates/funcmap.go]: Fixed with 329b3f1

**Review by hslatman — APPROVED** [review]

**Pull request merged**